### ISSUE TITLE
[BUG] Error Creating Detector [security_analytics_exception] normalizer [lowercase_keyword] not found for field

### ISSUE THREAD

**Issue by jsun64128** [opensearch-project/security-analytics#697]
**What is the bug?**
When trying to create a security analytics detector I get an error "[security_analytics_exception] normalizer [lowercase_keyword] not found for field 'x'". x being <Index-identifier>_<Unique string> Ex.  25-000873_J6yHbIsBON-lwPZfCF7R

**How can one reproduce the bug?**
1. Go to 'Security Analytics'
2. Click on 'Create Detector'
3. Configuration is for all elastic beats data sources "elastic_windows*"
4. Log Type 'windows'
5. Select all Rules
6. All mappings automatically applied except for 'windows-message' which was manually mapped to 'event.original'
7. Alert Trigger used default settings did not add channel during testing
8. Click 'Create Detector'
9. See error
10. "[security_analytics_exception] normalizer [lowercase_keyword] not found for field"

**What is the expected behavior?**
When creating indexes we already apply normalizers like lowercase_keyword. I pasted the code we utilize in the additional context section.

**What is your host/environment?**
 - OS: Windows 11
 - Version: OpenSearch 2.8
 - Plugins: 

**Do you have any screenshots?**
![Screenshot 2023-10-26 110448](https://github.com/opensearch-project/security-analytics/assets/149092256/487ced92-4d4e-4283-8c5c-c715be40c899)

**Do you have any additional context?**
When we create an index, we added the lowercase_keyword normalizer and set it as the normalizer for the message field
`{
  "mappings": {
    "properties": {
      "message": {
        "type": "keyword",
        "normalizer": "lowercase_keyword"
      }
    }
  },
  "settings": {
    "analysis": {
      "normalizer": {
        "lowercase_keyword": {
          "type": "custom",
          "filter": [
            "lowercase"
          ]
        }
      }
    }
  }
}`

**Comment by mvanderlee** [comment]
I think this is related to the fact that we don't have component templates but just regular index templates. 
As soon as we tried enabling a Detector, this project injected it's own component template and completely overwrote our own indices. 

This causes multiple issues:
1. Detector doesn't work
2. ISM policies no longer work, so our indices have been growing at a rapid pace without rolling over
3. Indices no longer have the correct mapping causing search errors. 

In short. attempting to enable Detectors completely broke out OS cluster. 

"You'll own nothing and be happy" comes to mind, even with our own data.

**Comment by mvanderlee** [comment]
Follow up, the broken indices and ISM policy were because use templates instead of index_templates. We now realize that templates were marked as legacy in 7.8. 
Still, it'd be great if an error would be thrown preventing the creation of overriding index_templates. The error should tell the user that in order to use Detectors, they will have to update to index_templates, preferably with a link to documentation or even a migration guide.

We have now updated all our templates to index_templates and fixed the ISM policies.

**Detectors still don't work**
I've tested on a dockerized version of 2.11.0 and still get the same error shown in the bug description.
Steps to reproduce
1. Start a new 2.11.0 cluster - `docker compose up` see attached docker-compose.yml file
2. Create index template 
    ```bash
    curl -X PUT -H 'Content-Type: application/json' --data "@schema.json" http://127.0.0.1:9200/_index_template/elastic_windows
    ```
3. Create index `<elastic_windows-{now/d}-000001>`
    ```bash
    curl -X PUT http://127.0.0.1:9200/%3Celastic_windows-%7Bnow%2Fd%7D-000001%3E
    ```
4. Create a Detector on the new index. Enable rule `Sticky Key Like Backdoor Usage` and for mapping select `winlog.message -> event.original`
5. Save and observe the error 
    ```
    [security_analytics_exception] normalizer [lowercase_keyword] not found for field [31-000001_uSNfhYsB4Y4-buH1DJSG]
    ```

**Comment by sbcd90** [comment]
hi @mvanderlee @jsun64128 , this is a known issue with using custom analyzers in doc-level monitors which the Security Analytics detectors internally create.
https://github.com/opensearch-project/alerting/issues/961
We are looking to fix this bug in future releases.

As a workaround for this, can you please try to use standard analyzers that OpenSearch support?

**Comment by mvanderlee** [comment]
@sbcd90 the alerting team just nonchalantly closed it and from the message I'm inferring a "Won't fix attitude" 
This would be a major shortcoming of OpenSearch. How will the security team handle this?

**Comment by duzvik** [comment]
Hit the same issue here.

**Comment by avidanov** [comment]
We encountered this issue with one of our customers. A possible workaround is to use analyzers instead of normalizers.

**Comment by thecodingshrimp** [comment]
** fix might belong in opensearch-project/alerting**

**How the bug works**

Security Analytics creates a custom query index (`.opensearch-sap-*-detectors-queries-optimized-*`) and passes it to the alerting plugin's `DocLevelMonitorQueries.kt` via `dataSources.queryIndex`. All field-mapping propagation happens inside alerting — SA owns none of it.

In `DocLevelMonitorQueries.kt`, the `leafNodeProcessor` closure (around line 329) does:
```kotlin
val newProps = props.toMutableMap()
```
This is a verbatim copy of every source field's attributes, including `analyzer`, `normalizer`, `similarity`, `search_analyzer`, and `search_quote_analyzer`. These attributes reference custom analysis objects defined in the source index's `settings.analysis.*` block. The query index is created from a fixed settings resource (`doc-level-queries.json`) that contains only `{ "index": { "mapping": { "total_fields": { "limit": 10000 } }, "hidden": true } }` — no `analysis` block. OpenSearch rejects the `PutMappingRequest` because the referenced analyzer/normalizer/similarity does not exist in the query index's settings.

**Why SA cannot fix this unilaterally**

SA has no hook to intercept `leafNodeProcessor`'s attribute copy. The `queryIndexMappingsByType` escape hatch in `DataSources.kt` is restricted by alerting's own validation to `text` fields + `analyzer` only — `normalizer` on `keyword` fields (the exact failure here) is excluded.

**The correct fix (strip analysis attributes in alerting)**

In `leafNodeProcessor`, strip `analyzer`, `normalizer`, `similarity`, `search_analyzer`, and `search_quote_analyzer` from the copied field attributes before they reach `PutMappingRequest`. These attributes are meaningless on the query index — the query index stores Percolator query documents and does not run user-facing text search; fields falling back to the default analyzer is correct for its purpose. This requires no index close/open cycle and does not conflict with the shared-query-index architecture.

The alternative (copying `settings.analysis.*` from the source index into the query index at creation time) is architecturally blocked: the query index is shared across all monitors, and conflicting custom analyzer/normalizer definitions from different source indices cannot be safely merged into one shared index.

**Broader scope (related: #1798)**

opensearch-project/security-analytics#1798 documents a broader variant of this failure covering `analyzer`, `search_analyzer`, `search_quote_analyzer`, `normalizer`, and `similarity` across multiple field types — not just `normalizer` on `keyword`. The root cause and fix location are identical.

**Upstream tracking**

This is the same defect as opensearch-project/alerting#961, which was reopened after the initial "won't fix" closure but remains unresolved as of OpenSearch v2.19.5. The strip-attributes fix approach directly invalidates the architectural objection raised there (which only applies to copying settings, not stripping attributes).

Until alerting#961 is fixed, the only workaround is to avoid custom analyzers, normalizers, and similarities on any field in a source index that SA monitors. Built-in analyzers (`standard`, `keyword`, `english`, etc.) work fine.

credit: claude
